### PR TITLE
Implement a couple more dataflow intrinsics

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
* `Type.GetNestedType` - should be straighforward
* `TypeDelegator..ctor` - this is used in CoreLib - we should just treat it as identity so that we don't lose dataflow annotations
* `object.GetType` - I specifically need the delegate part for `CoreLib`; the rest of the behavior is nice to have but I don't expect it to kick in too often.

Creating as draft because it obviously misses tests, but it's Friday...

### CODE REVIEW

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 1]
```diff
@@ -1483,6 +1606,10 @@ string GetDynamicallyAccessedMemberKindsDescription (DynamicallyAccessedMemberTy
 			return string.Join (" | ", results.Select (r => r.ToString ()));
 		}
 
+		static DynamicallyAccessedMemberTypes GetDynamicallyAccessedMemberTypesFromBindingFlagsForTypes (BindingFlags bindingFlags) =>
```
Comment (vitek-karas) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:1609]:
```suggestion
		static DynamicallyAccessedMemberTypes GetDynamicallyAccessedMemberTypesFromBindingFlagsForNestedTypes (BindingFlags bindingFlags) =>
```

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 2]
```diff
@@ -1083,7 +1190,7 @@ void RequireDynamicallyAccessedMembers (ref ReflectionPatternContext reflectionC
 			foreach (var uniqueValue in value.UniqueValues ()) {
 				if (uniqueValue is LeafValueWithDynamicallyAccessedMemberNode valueWithDynamicallyAccessedMember) {
 					if (!valueWithDynamicallyAccessedMember.DynamicallyAccessedMemberKinds.HasFlag (requiredMemberKinds)) {
-						reflectionContext.RecordUnrecognizedPattern ($"The {GetValueDescriptionForErrorMessage (valueWithDynamicallyAccessedMember)} " +
+						reflectionContext.RecordUnrecognizedPattern ($"{reflectionContext.SourceMethod} The {GetValueDescriptionForErrorMessage (valueWithDynamicallyAccessedMember)} " +
```
Comment (vitek-karas) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:1193]: This should be part of a different change - there are more places where we need to add something like this.

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 3]
```diff
@@ -1266,15 +1373,31 @@ void MarkFieldsOnTypeHierarchy (ref ReflectionPatternContext reflectionContext,
 			}
 		}
 
-		void MarkNestedTypesOnType (ref ReflectionPatternContext reflectionContext, TypeDefinition type, Func<TypeDefinition, bool> filter)
+		TypeDefinition[] MarkNestedTypesOnType (ref ReflectionPatternContext reflectionContext, TypeDefinition type, Func<TypeDefinition, bool> filter, BindingFlags bindingFlags = BindingFlags.Default)
```
Comment (vitek-karas) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:1376]: Please change the callsite in here https://github.com/mono/linker/pull/1195/files#diff-d634a3c96357eb6bf5b24034efb1e61aR1253-R1257 to use the BIndingFlags just like all the other kinds... (no need for the filter in those cases).

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs (hunk 2 above)
Comment (MichalStrehovsky) [reply src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs]: Oops, that was not intentional.

File: test/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs [hunk 4]
```diff
@@ -825,7 +825,7 @@ static string GetFullMemberNameFromDefinition (IMetadataTokenProvider member)
 				member = memberReference.Resolve ();
 
 			if (member is IMemberDefinition memberDefinition) {
-				if (memberDefinition is TypeDefinition) {
+				if (memberDefinition is TypeDefinition type) {
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs:828]: This seems to be unnecessary change...
